### PR TITLE
Jake/optimized knn

### PR DESCRIPTION
**Pull request by jake-ringdna**
Moved old model file (crude way to keep from losing info if something breaks on deployment).

New model parameters set using Bayesian optimization, with the following models and value ranges:
{
    'KNeighborsClassifier':{
        'n_neighbors':(2,10),
        'leaf_size':(1,50),
        'p':(1,2)
    },
    'RandomForestClassifier':{
        'max_features':['log2',None,'sqrt'],
        'n_estimators':(50,500),
        'min_samples_leaf':(1,50),
        'min_samples_split':(2,50)
    }
}

Best results were:

KNeighborsClassifier(leaf_size=33, n_neighbors=3, p=2)

New F1 Score: ~.24

Next iteration should include additional training data, more features, more optimization params, and better-optimized ranges.

### CODE REVIEW

**Comment by correlator** [comment]
I'd like to see the code you used to generate this model in the train method of the processor. That is the place to document how the model was trained. @jake-ringdna

**Comment by correlator** [comment]
https://github.com/blackbox-engineering/ring_pipeline/blob/master/nostradatas/processors/model_processors/demo_set_inference_processor.py#L29-L47

**Comment by correlator** [comment]
@jake-ringdna code has been updated so that model matches the model that is serialized in the wheel.  Merge if you're happy with it and it's accurate

**Comment by jake-ringdna** [comment]
Okay @correlator, requested changes have been made. Please take a look and let me know if this is where we want it for now.

File: nostradatas/pipelines/model_pipelines/demo_set_inference_pipeline.py [hunk 1]
```diff
@@ -21,7 +21,7 @@ def process(self):
             call.infered_demo_set = inference
             self.internal_employee_store.add_call(call.internal_user_id, call)
 
-    def train(self):
+    def quick_train(self):
```
Comment (correlator) [inline nostradatas/pipelines/model_pipelines/demo_set_inference_pipeline.py]: very nice!

File: nostradatas/pipelines/model_pipelines/demo_set_inference_pipeline.py [hunk 2]
```diff
@@ -32,7 +32,11 @@ def train(self):
         training_set = pd.concat([demo_sets, demo_not_sets])
 
         training_set = training_set.sample(frac=1)
-        self.demo_set_inference_processor.train(training_set)
+        self.demo_set_inference_processor.train(dataframe)
+
+    def train(self):
+        dataframe = self.load_dataframe_from_internal_store()
+        self.demo_set_inference_processor.train(dataframe)
```
Comment (correlator) [inline nostradatas/pipelines/model_pipelines/demo_set_inference_pipeline.py]: This won't be a balanced data set.

File: nostradatas/processors/model_processors/demo_set_inference_processor.py [hunk 3]
```diff
@@ -27,23 +28,94 @@ def process(self, dataframe: DataFrame) -> int:
...
+            'min_samples_split':(2,50)
+            }
+        }
+
+        def _model_opt(x, y, model, model_params, max_evals=100):
+            from hyperopt import hp, tpe, fmin, Trials
+            
+            uniform_params = {}
+            choice_params = {}
+            for key in model_params:
```
Comment (correlator) [inline nostradatas/processors/model_processors/demo_set_inference_processor.py]: isn't model_params a dictionary?  If so, let's be explicit about calling 
```
for key in model_params.keys():
```

could also save a lot of `[key]`s below by doing a 
```
for param_name, param_values in model_params.items()
```

File: nostradatas/processors/model_processors/demo_set_inference_processor.py [hunk 4]
```diff
@@ -27,23 +28,94 @@ def process(self, dataframe: DataFrame) -> int:
         return knn_predictor.predict(normalized_dataframe)
 
     def train(self, dataframe: DataFrame):
+        
+        param_playbook = {
```
Comment (correlator) [inline nostradatas/processors/model_processors/demo_set_inference_processor.py]: These are always referred to model params below, consider renaming to `model_param_lookup_dictionary` to save the reader from having to think.  My brain is the tiniest :)

File: nostradatas/processors/model_processors/demo_set_inference_processor.py [hunk 5]
```diff
@@ -27,23 +28,94 @@ def process(self, dataframe: DataFrame) -> int:
...
+                    uniform_params[key] = hp.quniform(*tuple(uniform_params[key]))
+                else:
+                    choice_params[key] = hp.choice(key,model_params[key])
+                    
+            adjusted_params = uniform_params
+            adjusted_params.update(choice_params)
+            
+            trials = Trials()
+            trials.delete_all()
+            def _sub_model_opt(model_params,x=x,y=y,model=model):
```
Comment (correlator) [inline nostradatas/processors/model_processors/demo_set_inference_processor.py]: is this a method in a method?  I'm starting to have trouble keeping track of the indentation here.  the good name is the only thing that saved me 👍

File: nostradatas/processors/model_processors/demo_set_inference_processor.py [hunk 6]
```diff
@@ -27,23 +28,94 @@ def process(self, dataframe: DataFrame) -> int:
...
+                    uniform_params[key].append(1)
+                    uniform_params[key] = hp.quniform(*tuple(uniform_params[key]))
+                else:
+                    choice_params[key] = hp.choice(key,model_params[key])
+                    
+            adjusted_params = uniform_params
+            adjusted_params.update(choice_params)
+            
+            trials = Trials()
+            trials.delete_all()
```
Comment (correlator) [inline nostradatas/processors/model_processors/demo_set_inference_processor.py]: add an extra line to separate the function definition.  Having blank lines between variable assignments and no gaps between  function definitions makes the code harder to scan an follow the story.

File: nostradatas/processors/model_processors/demo_set_inference_processor.py [hunk 7]
```diff
@@ -27,23 +28,94 @@ def process(self, dataframe: DataFrame) -> int:
...
+            def _sub_model_opt(model_params,x=x,y=y,model=model):
+                for key in model_params:
```
Comment (correlator) [inline nostradatas/processors/model_processors/demo_set_inference_processor.py]: see above comment on iterating over dicts

File: nostradatas/processors/model_processors/demo_set_inference_processor.py [hunk 8]
```diff
@@ -27,23 +28,94 @@ def process(self, dataframe: DataFrame) -> int:
...
+                    val = model_params[key]
```
Comment (correlator) [inline nostradatas/processors/model_processors/demo_set_inference_processor.py]: `model_param` is more descriptive and easier to follow than `val` . Imagine yourself coming back to this in 3 months and having to re-grok it :)

File: nostradatas/processors/model_processors/demo_set_inference_processor.py [hunk 9]
```diff
@@ -27,23 +28,94 @@ def process(self, dataframe: DataFrame) -> int:
...
+                model = eval(f'{model}(**model_params)')
+                f1 = cross_val_score(model, x, y, cv=5, scoring='f1', n_jobs=5).mean()
+                return -f1
+            
+            best = fmin(fn = _sub_model_opt,
+                        space = adjusted_params, algo=tpe.suggest, 
+                        max_evals = max_evals)
+            for key in best:
+                val = best[key]
+                if key in choice_params:
```
Comment (correlator) [inline nostradatas/processors/model_processors/demo_set_inference_processor.py]: I got stuck here trying to figure out what choice_params are.  They are categorical parameters right?  Once that aren't a continuous range?

File: nostradatas/processors/model_processors/demo_set_inference_processor.py [hunk 10]
```diff
@@ -27,23 +28,94 @@ def process(self, dataframe: DataFrame) -> int:
...
+            },
+        'RandomForestClassifier':{
+            'max_features':['log2',None,'sqrt'],
+            'n_estimators':(50,500),
+            'min_samples_leaf':(1,50),
+            'min_samples_split':(2,50)
+            }
+        }
+
+        def _model_opt(x, y, model, model_params, max_evals=100):
```
Comment (correlator) [inline nostradatas/processors/model_processors/demo_set_inference_processor.py]: This is fine for now, but I have a feeling that model optimization will be pulled out into it's own processor sub class at some ponit in the future so it can be used by many other processors.  No action needed now, just flagging it.

File: nostradatas/pipelines/model_pipelines/demo_set_inference_pipeline.py (hunk 2 above)
Comment (jake-ringdna) [reply nostradatas/pipelines/model_pipelines/demo_set_inference_pipeline.py]: True. Balancing is something we need to evaluate further. Given the size of this dataset, I've defaulted to avoiding it; I think we're going to need every training row we can get. Since we _are_ still getting relatively high false positives, I don't think this is going to help us with the current dataset. However, I went ahead and tried downsampling the majority class just in case. The results weren't good.

We may want to assess downsampling ratio as a tunable feature in the future. We may also want to look at bootstrapping techniques for the minority class. For now I'm going to skip implementing, I don't see a short-term gain from it.

Comment (jake-ringdna) [reply nostradatas/pipelines/model_pipelines/demo_set_inference_pipeline.py]: Just to add some specificity for future reference: Rebalancing dropped Kneighbors performance from F1 of ~.25 to about .02. If tuned just right, it was able to hit an F1 of .18 with Random Forest. As expected, it's reducing precision and improving recall. There's probably an inflection point somewhere where this becomes advantageous. As we assess more models and collect more data we'll definitely want to think about this as something we can tune.

Comment (correlator) [reply nostradatas/pipelines/model_pipelines/demo_set_inference_pipeline.py]: Definitely keep a spreadsheet with this info.  This is gold as we seek to understand the system better.  Awesome job 🚀

File: nostradatas/processors/model_processors/demo_set_inference_processor.py (hunk 5 above)
Comment (jake-ringdna) [reply nostradatas/processors/model_processors/demo_set_inference_processor.py]: Yeah, not my favorite way of doing this. Unfortunately hyperopt requires a loss function with a dictionary as an input and an output it seeks to minimize, which was a little too complicated for a lambda. Since this is a single-use function, nesting and "_"-prefix is the (ugly) solution I came up with.

File: nostradatas/processors/model_processors/demo_set_inference_processor.py (hunk 9 above)
Comment (jake-ringdna) [reply nostradatas/processors/model_processors/demo_set_inference_processor.py]: Yes, essentially discrete lists of values from which it's using random choice to make selections. It's kind of a crappy naming convention, but it follows the conventions used in hyperopt for distributions, which made it a little easier for me to keep track of.

File: nostradatas/processors/model_processors/demo_set_inference_processor.py [hunk 11]
```diff
@@ -27,23 +28,93 @@ def process(self, dataframe: DataFrame) -> int:
         return knn_predictor.predict(normalized_dataframe)
 
     def train(self, dataframe: DataFrame):
+        
+        model_param_lookup_dictionary = {
```
Comment (correlator) [inline nostradatas/processors/model_processors/demo_set_inference_processor.py]: 👍

File: nostradatas/processors/model_processors/demo_set_inference_processor.py [hunk 12]
```diff
@@ -27,23 +28,93 @@ def process(self, dataframe: DataFrame) -> int:
...
+        'RandomForestClassifier':{
+            'max_features':['log2',None,'sqrt'],
+            'n_estimators':(50,500),
+            'min_samples_leaf':(1,50),
+            'min_samples_split':(2,50)
+            }
+        }
+
+        def _model_opt(x, y, model, model_params, max_evals=100):
+            from hyperopt import hp, tpe, fmin, Trials
```
Comment (correlator) [inline nostradatas/processors/model_processors/demo_set_inference_processor.py]: This is a really cool lib

File: nostradatas/processors/model_processors/demo_set_inference_processor.py [hunk 13]
```diff
@@ -27,23 +28,93 @@ def process(self, dataframe: DataFrame) -> int:
...
+            def _sub_model_opt(model_params,x=x,y=y,model=model):
+                for param_name, param_value in model_params.items():
+                    if isinstance(param_value,float) and int(param_value)==param_value:
+                        model_params[param_name] = int(param_value)
+
+                model = eval(f'{model}(**model_params)')
+                f1 = cross_val_score(model, x, y, cv=5, scoring='f1', n_jobs=5).mean()
+                return -f1
+            
+            best = fmin(fn = _sub_model_opt,
```
Comment (correlator) [inline nostradatas/processors/model_processors/demo_set_inference_processor.py]: so does this fmin function return the index of the optimal params?
